server/coinlock/coinlocker.go

```diff
@@ -0,0 +1,237 @@
+	// defined as a map of OrderIDs to a CoinID slice since it is likely easiest
+	// for the caller to construct the input in this way.
+	LockCoins(orderCoins map[order.OrderID][]CoinID)
+}
+
+// MasterCoinLocker coordinates a book and swap coin locker. The lock status of
+// a coin may be checked, and the locker for the book and swapper may be
+// obtained via the Book and Swap methods.
+type MasterCoinLocker struct {
+	bookLock *AssetCoinLocker
+	swapLock *AssetCoinLocker
+}
+
+// NewMasterCoinLocker creates a new NewMasterCoinLocker.
+func NewMasterCoinLocker() *MasterCoinLocker {
+	return &MasterCoinLocker{
+		bookLock: NewAssetCoinLocker(),
+		swapLock: NewAssetCoinLocker(),
+	}
+}
+
+// CoinLocked indicates if a coin is locked by either the swap or book lock.
+func (cl *MasterCoinLocker) CoinLocked(coin CoinID) bool {
+	lockedBySwap := cl.swapLock.CoinLocked(coin)
+	if lockedBySwap {
+		return true
+	}
+
+	return cl.bookLock.CoinLocked(coin)
+}
+
+// OrderCoinsLocked lists all coins locked by a given order are locked by either
+// the swap or book lock.
+func (cl *MasterCoinLocker) OrderCoinsLocked(oid order.OrderID) []CoinID {
+	coins := cl.swapLock.OrderCoinsLocked(oid)
+	if len(coins) > 0 {
+		return coins
+	}
+
+	// TODO: Figure out how we will handle the evolving coinIDs for an order
+	// with partial fills and decide how to merge the results of both swap and
+	// book locks.
+	return cl.bookLock.OrderCoinsLocked(oid)
+}
+
+// Book provides the market-level CoinLocker.
+func (cl *MasterCoinLocker) Book() CoinLocker {
+	return &bookLocker{cl}
+}
+
+// Swap provides the swap-level CoinLocker.
+func (cl *MasterCoinLocker) Swap() CoinLocker {
+	return &swapLocker{cl}
+}
+
+type bookLocker struct {
+	*MasterCoinLocker
+}
+
+// LockOrdersCoins locks all coins for the given orders.
+func (bl *bookLocker) LockOrdersCoins(orders []order.Order) {
+	bl.bookLock.LockOrdersCoins(orders)
+}
+
+// LockOrdersCoins locks coins associated with certain orders.
+func (bl *bookLocker) LockCoins(orderCoins map[order.OrderID][]CoinID) {
+	bl.bookLock.LockCoins(orderCoins)
+}
+
+// UnlockOrderCoins unlocks all locked coins associated with an order.
+func (bl *bookLocker) UnlockOrderCoins(oid order.OrderID) {
+	bl.bookLock.UnlockOrderCoins(oid)
+}
+
+var _ (CoinLocker) = (*bookLocker)(nil)
+
+type swapLocker struct {
+	*MasterCoinLocker
+}
+
+// LockOrdersCoins locks all coins for the given orders.
+func (bl *swapLocker) LockOrdersCoins(orders []order.Order) {
+	bl.swapLock.LockOrdersCoins(orders)
+}
+
+// LockOrdersCoins locks coins associated with certain orders.
+func (bl *swapLocker) LockCoins(orderCoins map[order.OrderID][]CoinID) {
+	bl.swapLock.LockCoins(orderCoins)
+}
+
+// UnlockOrderCoins unlocks all locked coins associated with an order.
+func (bl *swapLocker) UnlockOrderCoins(oid order.OrderID) {
+	bl.swapLock.UnlockOrderCoins(oid)
+}
+
+var _ (CoinLocker) = (*swapLocker)(nil)
+
+type coinIDKey string
```

Comment:
This is kinda icky, so I'm open to other ideas for a map key, although since there's no restriction on CoinID length in bytes, I'm at a loss for ideas.

Reply:
I'm okay with strings. I know they take a performance hit in map look ups, but I'm guessing it's not significant enough to warrant the complexity of extra IDs.